qutip/core/data/norm.pyx

```diff
@@ -212,6 +229,7 @@ trace.__doc__ =\
     """
 trace.add_specialisations([
     (CSR, trace_csr),
+    (Dense, trace_dense),
```

Comment:
`trace_dense` signature does not have the `sparse, tol, maxiter` parameters of the trace dispatcher so:
`qt.data.norm.trace(qt.qeye(5).to('dense').data)` fails: 
```TypeError: qutip.core.data.norm.trace_dense() takes exactly one argument (4 given)```

The dispatcher signature must only contain parameters that all specializations accept.

Reply:
I did not know that `trace_norm` had to take those arguments necessarily. Sorry for that. What do you thing about adding a check in the `add_specialisaiton` method to check whether the added specialisations accept the expected arguments (?). 

I any case, these arguments dont seem to make sense for dense_norm. 
I am not sure what to do here. I can revert the changes and not include the `trace_dense` in this PR. However, it does seem that a trace_dense is useful as you would like to avoid the conversion to CSR when doing the `norm.trace`. Mainly because, by default, you convert back to dense...

Reply:
I suggest we remove the `sparse` parameter and `tol` and `maxiter` from the dispatcher.
`_trace_csr` would become the specialization for sparse. For a user to change the `tol`, he would need to call the specialization directly, but since it does not make sense for other data type, it should be fine.

Reply:
I was worried that this would break backwards compatibility but I also agree that removing those parameters is probably necessary. Should I proceed then?

Reply:
Backwards compatibility could be fixed in `Qobj.norm`. But I don't believe that these options are that commonly used.